src/Security/Security.php

```diff
@@ -303,12 +392,8 @@ public static function permissionFailure($controller = null, $messageSet = null)
                 $message = $messageSet['default'];
             }
 
-            // Somewhat hackish way to render a login form with an error message.
-            $me = new Security();
-            $form = $me->LoginForm();
-            $form->sessionMessage($message, ValidationResult::TYPE_WARNING);
-            Session::set('MemberLoginForm.force_message', 1);
-            $loginResponse = $me->login();
+            Security::setLoginMessage($message, ValidationResult::TYPE_WARNING);
+            $loginResponse = (new Security())->login(new HTTPRequest('GET', '/'));
```

Comment:
OK, so answering the question briefly. Two reasons:

 * The only way to reconstruct a view is to get a controller object and call an action handler on it. We don't have the complete login view available as any object other than this request handler.
 * When using a custom request handler, it requires an HTTPRequest object as an argument. I'm not sure what happens if you pass `null` but probably annoying things.